client/mysqldump.c

```diff
@@ -4165,6 +4252,397 @@ static char *getTableName(int reset)
+      "       WITH_ADMIN_OPTION "
+      "FROM create_role_order "
+      "ORDER BY n,"
+      "         FROM_USER,"
+      "         FROM_HOST,"
+      "         TO_USER,"
+      "         TO_HOST,"
+      "         WITH_ADMIN_OPTION"))
+    return 1;
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    /* MySQL-8.0 export capability */
+    if (opt_replace_into)
+      fprintf(md_result_file,
+        "/*!80001 DROP ROLE IF EXISTS %s */;\n", row[0]);
+    fprintf(md_result_file,
+      "/*!80001 CREATE ROLE %s%s */;\n", opt_ignore ? "IF NOT EXISTS " : "", row[0]);
+    /* By default created with current role */
+    fprintf(md_result_file,
+      "%sROLE %s%s WITH ADMIN mariadb_dump_import_role */;\n",
+      opt_replace_into ? "/*M!100103 CREATE OR REPLACE ": "/*M!100005 CREATE ",
+      opt_ignore ? "IF NOT EXISTS " : "", row[0]);
+    fprintf(md_result_file, "/*M!100005 GRANT %s TO %s%s*/;\n",
+            row[0], row[1], (row[2][0] == 'Y') ? " WITH ADMIN OPTION " : "");
+  }
+  mysql_free_result(tableres);
+
+  /* users and their default role */
+  if (maria_roles_exist && mysql_query_with_error_report(mysql, &tableres,
+      "select IF(default_role='', 'NONE', QUOTE(default_role)) as r,"
+      "concat(QUOTE(User), '@', QUOTE(Host)) as u FROM mysql.user  "
+      "/*M!100005 WHERE is_role='N' */"))
+    return 1;
+  if (mysql_roles_exist && mysql_roles_exist && mysql_query_with_error_report(mysql, &tableres,
+      "SELECT IF(DEFAULT_ROLE_HOST IS NULL, 'NONE', CONCAT(QUOTE(DEFAULT_ROLE_USER), '@', QUOTE(DEFAULT_ROLE_HOST))) as r,"
+      "  CONCAT(QUOTE(mu.USER),'@',QUOTE(mu.HOST)) as u "
+      "FROM mysql.user mu LEFT JOIN mysql.default_roles using (USER, HOST)"))
+    return 1;
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    if (dump_grants(row[1]))
+      result= 1;
+    fprintf(md_result_file, "/*M!100005 SET DEFAULT ROLE %s FOR %s */;\n", row[0], row[1]);
+    fprintf(md_result_file, "/*!80001 ALTER USER %s DEFAULT ROLE %s */;\n", row[1], row[0]);
+  }
+  mysql_free_result(tableres);
+
+  if (maria_roles_exist && mysql_query_with_error_report(mysql, &tableres,
+      "SELECT DISTINCT QUOTE(m.role) AS r "
+      "   FROM mysql.roles_mapping m"
+      "   JOIN mysql.user u ON u.user = m.role"
+      "   WHERE is_role='Y'"
+      "     AND Admin_option='Y'"
+      "   ORDER BY m.role"))
+    return 1;
+  if (mysql_roles_exist && mysql_query_with_error_report(mysql, &tableres,
+      "SELECT DISTINCT CONCAT(QUOTE(FROM_USER),'@', QUOTE(FROM_HOST)) AS r "
+      "FROM mysql.role_edges"))
+    return 1;
+  while ((row= mysql_fetch_row(tableres)))
+  {
+    if (dump_grants(row[0]))
+      result= 1;
+  }
+  mysql_free_result(tableres);
+  /* switch back */
+  fputs("SET ROLE NONE;\n"
+        "DROP ROLE mariadb_dump_import_role;\n"
+        "/*M!100203 EXECUTE IMMEDIATE CONCAT('SET ROLE ', @current_role) */;\n",
```

Comment:
I skipped role dumping, since don't know too much about algorithm/use cases how they are used. 
Please explain more in jira comments with valid examples and different scenarios.

Reply:
case for switching role per above.

Yes role dumping is hard. Hierarchies and an output that need to be functional made this complicated (just like sequences). Do you think a better explanation is needed on the role dumping? If so comments or commit message are for this, not JIRA.